lib/pushmi_pullyu/cli.rb

```diff
@@ -32,39 +32,37 @@ def run
 
   def start_server
     setup_signal_traps
+
     setup_log
     print_banner
 
-    begin
-      run_tick_loop
-    rescue Interrupt
-      logger.info 'Shutting down'
-      @running = false
-      logger.info 'Bye!'
-      exit(0)
-    end
+    run_tick_loop
   end
 
   private
 
   def setup_signal_traps
-    Signal.trap('INT') { raise Interrupt }
-    Signal.trap('TERM') { raise Interrupt }
+    Signal.trap('INT') { Thread.new { shutdown } }
+    Signal.trap('TERM') { Thread.new { shutdown } }
     Signal.trap('HUP') do
-      if config.logfile
-        logger.debug 'Received SIGHUP, reopening log file'
-        # TODO: reopen logs
-        # PushmiPullyu::Logging.reopen_logs(config.logfile)
+      Thread.new do
```

Comment:
The Thread.new stuff is ulgy... but since Ruby 2+,  Mutex#lock cannot live within a signal handler trap context. This prevents Logger from working since it uses an underlying mutex.

More info here:
https://bugs.ruby-lang.org/issues/7917

So this is kinda of a hack.

Alternative solutions here:
https://gist.github.com/mvidner/bf12a0b3c662ca6a5784

Seen a few projects use the The self-pipe trick? Maybe we should do one of these solutions.

The shutdown ones can avoid using logger by using `$stderr.puts` or something. But the reopening of log is unavoidable.